lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -237,10 +237,20 @@ static bool diagnoseUnsupportedControlFlow(ADContext &context,
 /// derivative generic signature (containing requirements), and substitution
 /// map. Returns true if error is emitted.
 static bool diagnoseUnsatisfiedRequirements(ADContext &context,
+                                            CanSILFunctionType origFnTy,
                                             GenericSignature derivativeGenSig,
                                             SubstitutionMap substMap,
                                             DifferentiationInvoker invoker,
                                             SourceLoc loc) {
+  // If the original function is polymorphic and its generic signature is the
+  // same as the derivative generic signature, then the requirements are
+  // satisfied. This check is necessary because the subsequent logic does not
+  // correctly handle polymorphic original functions.
+  // TODO(TF-1055): Can be removed after we have a robust solution for TF-1055.
```

Comment:
I wonder what's the specific symptom of TF-1055?

Referencing the example:
```
sil_differentiability_witness @foo <T: A> { ... }

sil_differentiability_witness @foo <T: B> { ... }

sil @foo : <T> (T) -> T

sil @example
bb0:
  %1 = function_ref @foo : <T> (T) -> T
  %2 = differentiable_function %1
  %3 = differentiable_function_extract [vjp] %2
  ...
```
Currently, is the symptom that one differentiability witness is chosen at random during `differentiable_function` canonicalization?

Reply:
You'd get an error that foo is not differentiable, even if the next instruction applies the vjp to something that conforms to `A`. However, I don't know if any swift code will actually generate SIL like that.

With this check here, I know of no swift code that causes any symptoms.

A symptom that you can get from real swift, if you remove the check here, is:
```
protocol P {
  @differentiable
  func f(_ x: Float) -> Float
}

extension P {
  @differentiable
  func f(_ x: Float) -> Float
}

struct S: P {}
```
=>
```
error: function is not differentiable because `Self: P` is not satisfied
```
(error from memory, don't have a compiler to test it on right now)

Reply:
The reason that error happens is that the differentiability witness table for `P.f` has a `Self: P` constraint, and the substitution map that gets passed in to `diagnoseUnsatisfiedRequirements` doesn't have anything satisfying those constraints.

Reply:
I see, thanks for the details! I wonder if this is a deficiency in the code around `diagnoseUnsatisfiedRequirements` (e.g. logic for computing the substitution map). I'll try to reproduce the issue after this patch lands.